File: data-jpa/src/main/java/top/bettercode/summer/data/jpa/query/mybatis/MybatisResultTransformer.java
package top.bettercode.summer.data.jpa.query.mybatis;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.List;
import org.apache.ibatis.mapping.MappedStatement;
import org.hibernate.jpa.spi.NativeQueryTupleTransformer;

/**
 * @author Peter Wu
 */
public class MybatisResultTransformer extends NativeQueryTupleTransformer {

  private static final long serialVersionUID = 1L;

  private final MappedStatement mappedStatement;

  public MybatisResultTransformer(MappedStatement mappedStatement) {
    this.mappedStatement = mappedStatement;
  }


  public Object transform(ResultSet resultSet) {
    return this.transformList(resultSet, 1).get(0);
  }

  public List<?> transformList(ResultSet resultSet) {
    try {
      return transformList(resultSet, resultSet.getFetchSize());
    } catch (SQLException e) {
      throw new RuntimeException(e);
    }
  }

  public List<?> transformList(ResultSet resultSet, int maxRows) {
    try {
      if (maxRows == 0) {
        maxRows = Integer.MAX_VALUE;
      }
      return new MybatisResultSetHandler(mappedStatement).handleResultSets(resultSet, maxRows);
    } catch (SQLException e) {
      throw new RuntimeException(e);
    }
  }

}

File: data-jpa/src/main/kotlin/top/bettercode/summer/data/jpa/query/mybatis/MybatisResultTransformer.kt
package top.bettercode.summer.data.jpa.query.mybatis

import org.apache.ibatis.mapping.MappedStatement
import org.hibernate.transform.ResultTransformer
import java.sql.ResultSet
import javax.persistence.Tuple

/**
 * @author Peter Wu
 */
class MybatisResultTransformer(
    private val mappedStatement: MappedStatement,
    private val isStreamQuery: Boolean
) : ResultTransformer {

    private var resultSetMetaData: TupleResultSetMetaData? = null
    var autoCloseResultSet = true

    override fun transformTuple(tuple: Array<Any?>, aliases: Array<String>): Any {
        val tupleImpl = NativeTupleImpl(tuple, aliases)
        if (isStreamQuery) {
            val tuples = listOf(tupleImpl)
            val resultSetMetaData: TupleResultSetMetaData
            if (this.resultSetMetaData == null) {
                synchronized(this) {
                    if (this.resultSetMetaData == null) {
                        resultSetMetaData = TupleResultSetMetaData(tuples)
                        if (resultSetMetaData.complete) {
                            this.resultSetMetaData = resultSetMetaData
                        }
                    } else {
                        resultSetMetaData = this.resultSetMetaData!!
                    }
                }
            } else {
                resultSetMetaData = this.resultSetMetaData!!
            }
            val resultSet = TupleResultSet(tuples, resultSetMetaData)
            return MybatisResultSetHandler(mappedStatement, autoCloseResultSet).handleResultSets(
                resultSet,
                Int.MAX_VALUE
            ).first()!!
        } else {
            return tupleImpl
        }
    }

    @Suppress("UNCHECKED_CAST")
    override fun transformList(list: List<*>): List<Any?> {
        val tuples = list as List<Tuple>
        val resultSetMetaData: TupleResultSetMetaData
        if (this.resultSetMetaData == null) {
            synchronized(this) {
                if (this.resultSetMetaData == null) {
                    resultSetMetaData = TupleResultSetMetaData(tuples)
                    if (resultSetMetaData.complete) {
                        this.resultSetMetaData = resultSetMetaData
                    }
                } else {
                    resultSetMetaData = this.resultSetMetaData!!
                }
            }
        } else {
            resultSetMetaData = this.resultSetMetaData!!
        }
        val resultSet = TupleResultSet(tuples, resultSetMetaData)
        return MybatisResultSetHandler(mappedStatement, autoCloseResultSet).handleResultSets(
            resultSet,
            Int.MAX_VALUE
        )
    }

    fun transformResultSet(resultSet: ResultSet?): Any? {
        return this.transformListResultSet(resultSet, 1)[0]
    }

    fun transformListResultSet(resultSet: ResultSet): List<Any?> {
        return transformListResultSet(resultSet, resultSet.fetchSize)
    }

    fun transformListResultSet(resultSet: ResultSet?, maxRows: Int): List<Any?> {
        var maxRows1 = maxRows
        if (maxRows1 == 0) {
            maxRows1 = Int.MAX_VALUE
        }
        return MybatisResultSetHandler(mappedStatement, autoCloseResultSet).handleResultSets(
            resultSet,
            maxRows1
        )
    }

}
